File: delta_main.py
# -*- coding: utf-8 -*-
"""
Created on Thu Sep 01 20:28:52 2016

@author: Nicholas
"""
import delta_synth_functions as d_synth
import numpy as np
import matplotlib.pyplot as plt
#import scipy.interpolate as interp
#import scipy.signal as sig

# model parameters for the synthetic data set
N_STATES = 3 # number of states
MEMORY = 5 # memory (in terms of # states)
V_SCALE = 1 #number of time steps per state step (assume state dwell ~1min)
TIME = 600 # time steps in sequence (assume time res ~10s and set series length ~20min)
TIME_SERIES = np.arange(0, TIME+V_SCALE, 1) #Time vector
TRANSITION_PROBS = np.array([[.3, .3, .5], \
                             [.2, .3, .1], \
                             [.5, .4, .4]])
#TRANSITION_PROBS = np.array([[0.7, 0.3, 0.2], [0.2, 0.5, 0.4], [0.1, 0.2, 0.4]]) # transition matrix
EMISSIONS = np.array([0, 4, 8], dtype = float) # emission values
V_NOISE = 0.4 # gaussian noise added to underlying states
#noise signal vector. R1 gives signal period in time steps, R2 gives magnitude relative to max(c)
EXP_NOISE = np.array([[2.], [.001]], dtype=float)  
PI0 = np.array([0, 1, 0], dtype = float)

##Call function to generate synthetic data
#synthetic (t_steps, v_scale, K, w, A, v, v_noise, exp_noise)
fluo =d_synth.synthetic(TIME, V_SCALE, N_STATES, MEMORY, TRANSITION_PROBS, EMISSIONS, V_NOISE, EXP_NOISE) 
fluo_zero = np.append(np.zeros(V_SCALE),fluo)
'''
    Try simplest approach first: take differences directly.
    Assess results using line plots + histogram
    
    Note: unassisted differentiation method begins to fail as noise approaches
          1/10 the separation of states.
'''
##take differences
diff = np.append(np.diff(fluo_zero),0)

#use knowledge of memory to disentangle loading rate from unloading rate
mem = MEMORY*V_SCALE

l_rates = np.zeros(len(diff), dtype = float)
l_rates[0:mem] = diff[0:mem]
#print l_rates
#iteratively calcualte loading rates
for i in range(mem,len(l_rates)):
    u = l_rates[i-mem]
    l_rates[i] = max(u,0) + diff[i] #assume l_rates < 0 are erroneous

#print fit
#plt.plot(TIME_SERIES, fluo_zero, TIME_SERIES, l_rates, TIME_SERIES, diff, linewidth=2.0) 
#histogram
plt.hist(l_rates)

File: MATLAB/Delta_main.m
%%Main file for loading rate inference project
%First call synthetic data function
A = [.2,.3,.2;.4,.6,.7;.4,.1,.1];
V = [0,2,4];
E_NOISE = [0,0];
E_FREQ = [5,1];
NOISE = 0;
SEQ = 50;
MEM = 5;
RES = 1;
%{
Args:
        seq_length (int): # of promoter states in sequence
        K (int): number of naive states
        res (int): number of discrete time steps per promoter state 
        w (int): memory
        A (stochastic float matrix): transition matrix
        v (float array): emission values
        v_noise (float): Scale of STD of Gaussian noise relative to 
        average state sep. Imitates inherent fluctuations in emission
        values
        exp_freq (int array): really the period of noise signals in time
        units
        exp_noise (float array): magnitude of gaussian noise signals
        relative to state separation
%}    

[fluo, fluo_interp, compound, naive] = Synthetic (SEQ, 3, RES, MEM, A, V, NOISE, E_FREQ, E_NOISE);
fluo_interp = horzcat(0,fluo_interp);
%zero negative terms
fluo_interp(fluo_interp<0) = 0;

t_interp = linspace(1,length(fluo), length(fluo_interp));

plot(1:SEQ, compound, 1:SEQ, fluo, t_interp, fluo_interp);
%%
length(fluo_interp)
%% Now take simple-minded differences approach (no filtering/smoothing)
%method yields unsatisfactory results, even when only noise is same freq
%as signal (i.e. state duration). Even worse when higher freq noise is added 
f_diffs = horzcat(0,diff(fluo_interp)).*RES;
f_2d = horzcat(0,diff(f_diffs)).*RES;
t_diffs = horzcat(0,t_interp(2:end));

%infer loading rates using diffs and mem
%l_rates = loading_rates(f_diffs, MEM, RES);
plot(1:SEQ, naive)
% histogram(l_rates)
%% Next try applying Savitzky-Golay filter before differntiating
W = 11; %width of interpolation window in t steps
HALF_W = (W-1)/2;
DEGREE = 9; %degree of poynomial to be used

%use sgolay to create filter of desired order (k) 
fluo_filt = sgolayfilt(fluo_interp,DEGREE,W);
%differentiate
smf_diffs = horzcat(0,diff(fluo_filt)).*RES;

%infer loading rates using diffs and mem
%sml_rates = loading_rates(smf_diffs, MEM, RES);

plot(1:SEQ, naive, t_interp, sml_rates)
%histogram(sml_rates)
% hold on
% histogram(naive)
